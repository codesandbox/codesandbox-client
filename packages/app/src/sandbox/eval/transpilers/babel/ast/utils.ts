import * as meriyah from 'meriyah';
import * as astring from 'astring';
import { customGenerator } from './generator';

export interface ESTreeAST {
  isDirty: boolean;
  program: meriyah.ESTree.Program;
}

export function parseModule(code: string): ESTreeAST {
  return {
    isDirty: false,
    program: meriyah.parseModule(code, {
      next: true,
      raw: true,
      jsx: true,
    }),
  };
}

export function generateCode(ast: ESTreeAST) {
  const finalCode = astring.generate(ast.program as any, {
    generator: customGenerator,
  });

  return `"use strict";\n${finalCode}`;
}
