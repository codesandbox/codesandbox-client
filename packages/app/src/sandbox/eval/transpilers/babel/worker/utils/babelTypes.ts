let types: unknown = null;

/**
 * Load babel types if the fs is needed, it greatly speeds up compilation of
 * custom plugins and presets, but we don't want to include it in the main bundle.
 */
export function loadBabelTypes(): Promise<void> {
  return import('@babel/types').then(typesResult => {
    types = typesResult;
  });
}

export function getBabelTypes() {
  if (!types) {
    throw new Error("Babel types aren't loaded");
  }

  return types;
}
