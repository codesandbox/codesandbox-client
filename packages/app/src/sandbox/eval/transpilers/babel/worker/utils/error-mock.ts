/* eslint-disable no-extend-native, no-useless-computed-key */
const getFileName = () => '/index.js';
const stackSplit = { getFileName };

let hasStack = false;

function checkHasStack() {
  if (hasStack) {
    return;
  }

  const origPrepareStackTrace = Error.prepareStackTrace;

  // eslint-disable-next-line
  Error.prepareStackTrace = function (_, stack) {
    return stack;
  };

  const stack = new Error().stack;
  Error.prepareStackTrace = origPrepareStackTrace;

  // @ts-ignore
  hasStack = typeof stack[0].getFileName === 'function';
}

checkHasStack();

class NewError extends Error {
  constructor(n: string) {
    super(n);

    // @ts-ignore
    this.stack = {
      [2]: stackSplit,
      split() {
        return [stackSplit, stackSplit, stackSplit, stackSplit];
      },
      replace() {
        return '';
      },
    };
  }
}

/**
 * Babel expects to be able to do this:
 *
 * ```js
 * const error = new Error();
 * const fileName = error.stack.split()[2].getFileName();
 * ```
 *
 * This works, but only in the V8 engine, so Firefox and Safari will break.
 * This function will hardcode the functionality.
 */
export function installErrorMock() {
  if (!hasStack) {
    // @ts-ignore
    self.Error = NewError;

    hasStack = true;
  }
}
