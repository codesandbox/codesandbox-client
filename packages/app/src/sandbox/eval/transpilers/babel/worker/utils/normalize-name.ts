export function normalizePluginName(name: string) {
  return name.replace('babel-plugin-', '').replace('@babel/plugin-', '');
}

export function normalizePresetName(name: string) {
  return name.replace('babel-preset-', '').replace('@babel/preset-', '');
}
