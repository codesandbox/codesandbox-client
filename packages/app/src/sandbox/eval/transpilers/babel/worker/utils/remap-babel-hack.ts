export function remapBabelHack() {
  if (self.module.exports.availablePlugins) {
    // Okay, Babel thought it was in a commonjs env and now put everything on
    // self.exports. We need to remap it.

    // @ts-ignore
    self.Babel = self.module.exports;

    // Reset module for next time Babel loads
    // @ts-ignore
    self.module = { exports: {} };
    const { exports } = self.module;
    self.exports = exports;
  }
}
