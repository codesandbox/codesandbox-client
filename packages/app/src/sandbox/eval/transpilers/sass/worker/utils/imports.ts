const lineRegex = /@import\s*['|"|`]([^"|'|`]*)['|"|`]/;

export function getImports(code: string): string[] {
  const lines = code.split('\n');
  const result = new Set();
  lines.forEach(line => {
    const match = line.match(lineRegex);
    if (match && match[1]) {
      result.add(match[1]);
    }
  });

  return Array.from(result) as string[];
}
