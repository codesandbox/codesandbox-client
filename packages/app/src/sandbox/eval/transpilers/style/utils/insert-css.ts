const wrapper = (id: string, css: string, webpackHMREnabled = false) => `
function createStyleNode(id, content) {
  var styleNode =
    document.getElementById(id) || document.createElement('style');

  styleNode.setAttribute('id', id);
  styleNode.type = 'text/css';
  if (styleNode.styleSheet) {
    styleNode.styleSheet.cssText = content;
  } else {
    styleNode.innerHTML = '';
    styleNode.appendChild(document.createTextNode(content));
  }
  document.head.appendChild(styleNode);
}

createStyleNode(
  ${JSON.stringify(id)},
  ${JSON.stringify(css)}
);

${webpackHMREnabled ? 'module.hot.accept()' : ''}
`;

export default function (
  id: string,
  css?: string,
  webpackHMREnabled?: boolean
) {
  const result = wrapper(id, css || '', webpackHMREnabled);
  return result;
}
