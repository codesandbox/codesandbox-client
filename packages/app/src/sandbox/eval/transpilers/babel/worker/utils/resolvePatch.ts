import * as resolve from 'browser-resolve';
import { packageFilter } from 'sandbox/eval/utils/resolve-utils';

/**
 * Patch 'resolve' to configure it to resolve esmodules by default. babel-plugin-macros goes
 * for the commonjs version, which we don't download anymore by default.
 */
export function patchedResolve() {
  const handler = {
    get(target, prop) {
      if (prop === 'sync') {
        return (p, options) => target.sync(p, { ...options, packageFilter });
      }

      return target[prop];
    },
  };
  return new Proxy(resolve, handler);
}
