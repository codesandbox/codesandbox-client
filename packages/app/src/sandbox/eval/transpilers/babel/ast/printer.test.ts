import { generateCode, parseModule } from './utils';

function processCode(code: string): string {
  return generateCode(parseModule(code));
}

describe('printer issues', () => {
  it('can convert + +', () => {
    const code = `
      c = (10.0, + +(15))
      `;

    expect(processCode(code)).toMatchSnapshot();
  });

  it('can convert -(--i)', () => {
    const code = `a = -(--i)`;
    expect(processCode(code)).toBe('"use strict";\na = - --i;\n');
  });

  it('can convert unicode line breaks', () => {
    const code = `const a = "[\\u2028]";`;
    expect(processCode(code)).toBe('"use strict";\nconst a = "[\\u2028]";\n');
  });
});
