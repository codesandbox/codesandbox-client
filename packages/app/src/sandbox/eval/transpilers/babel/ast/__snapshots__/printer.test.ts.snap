// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`printer issues can convert + + 1`] = `
""use strict";
c = (10.0, + + 15);
"
`;
